python/triton/config.py

```diff
@@ -0,0 +1,310 @@
+    store_binary_only: env_bool = env_bool("TRITON_STORE_BINARY_ONLY")
+    always_compile: env_bool = env_bool("TRITON_ALWAYS_COMPILE")
+    use_ir_loc: env_bool = env_bool("USE_IR_LOC")
+    enable_asan: env_bool = env_bool("TRITON_ENABLE_ASAN")
+    disable_line_info: env_bool = env_bool("TRITON_DISABLE_LINE_INFO")
+    frontend_debugging: env_bool = env_bool("TRITON_FRONT_END_DEBUGGING")
+    allow_non_constexpr_globals: env_bool = env_bool("TRITON_ALLOW_NON_CONSTEXPR_GLOBALS")
+
+
+class autotuning:
+    cache: env_bool = env_bool("TRITON_CACHE_AUTOTUNING")
+    print: env_bool = env_bool("TRITON_PRINT_AUTOTUNING")
+
+
+class LaunchHook(Protocol):
+
+    def __call__(self, metadata: LazyDict) -> None:
+        ...
+
+
+# This is of the form [attr_name, attr_val]
+# TODO: Use tuple instead of list for better typing.
+KernelAttr = list[str | int]
+
+
+class JitHookCompileInfo(TypedDict):
+    key: str
+    signature: dict[KernelParam, str]
+    device: int
+    constants: None
+    num_warps: int
+    num_ctas: int
+    num_stages: int
+    enable_fp_fusion: bool
+    launch_cooperative_grid: bool
+    extern_libs: tuple[tuple[str, str], ...]
+    configs: list[dict[tuple[int, ...], list[KernelAttr]]]
+    specialization_data: str
+    is_warmup: bool
+
+
+class JitHook(Protocol):
+
+    def __call__(self, *, key: str, repr: str, fn: JitFunctionInfo, compile: JitHookCompileInfo, is_manual_warmup: bool,
+                 already_compiled: bool) -> bool | None:
+        ...
+
+
+class runtime:
+    interpret: env_bool = env_bool("TRITON_INTERPRET")
+    debug: env_bool = env_bool("TRITON_DEBUG")
+    override_arch: env_str = env_str("TRITON_OVERRIDE_ARCH")
+
+    launch_enter_hook: LaunchHook | None = None
+    launch_exit_hook: LaunchHook | None = None
+
+    # Hook for inspecting compiled functions and modules
+    jit_cache_hook: JitHook | None = None
+    # Hook to signal that a kernel is done compiling and inspect compiled function.
+    # jit_cache_hook will always be called before compilation and jit_post_compile_hook after.
+    jit_post_compile_hook: JitHook | None = None
+
+
+class language:
+    fp32_default: env_str = env_str("TRITON_F32_DEFAULT")
+    default_fp_fusion: env_bool = env_bool("TRITON_DEFAULT_FP_FUSION", True)
+
+
+class nvidia:
+    cuobjdump: NvidiaTool = get_nvidia_tool("cuobjdump")  #field(default_factory=lambda: _get_nvidia_tool("cuobjdump"))
+    nvdisasm: NvidiaTool = get_nvidia_tool("nvdisasm")  #field(default_factory=lambda: _get_nvidia_tool("nvdisasm"))
+    ptxas: NvidiaTool = get_nvidia_tool("ptxas")  #field(default_factory=lambda: _get_nvidia_tool("ptxas"))
+
+    dump_nvptx: env_bool = env_bool("NVPTX_ENABLE_DUMP")
+    disable_ptxas_opt: env_bool = env_bool("DISABLE_PTXAS_OPT")
+    mock_ptx_version: env_str = env_str("TRITON_MOCK_PTX_VERSION")
+
+    libdevice_path: env_str = env_str("TRITON_LIBDEVICE_PATH")
+    libcuda_path: env_str = env_str("TRITON_LIBCUDA_PATH")
+
+
+class amd:
```

Comment:
Not sure if this is in scope for this PR, but what is the process for an out of tree backend for Triton? Is that an environment variable free for all or is there plan for an external hook you can plug into.